keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -233,7 +233,585 @@ yet but we think may be interesting to consider in the future.
+identifies a string attribute of the authenticated request and
+optionally includes a transformation.  The available string attributes
+are (1) namespace of a resource-oriented request (available only if
+the predicate accepts only resource-oriented requests) and (2)
+username.  If no transformation is indicated then the flow
+distinguisher is simply the selected request attribute.  There is only
+one transformation available, and it is based on a regex that is
+configured in the flow schema and contains a capturing group.  The
+transformation consists of doing a complete match against the regex
+and extracting submatch number 1; if the selected string does not
+match the regex then the transformation yields the empty string.
+
+### Assignment to a Queue
+
+A non-exempt RequestPriority object also has a number of queues (we
+are talking about a number here, not the actual set of queues; the
+queues exist independently at each apiserver).  If the
+RequestPriority’s number of queues is more than one then the following
+logic is used to assign a request to a queue.
+
+For a given priority at a given apiserver, each queue is identified by
+a numeric index (starting at zero).  A RequestPriority has a hand size
+H (so called because the technique here is an application of shuffle
+sharding), a small positive number.  When a request arrives at an
+apiserver the request flow identifier’s string pair is hashed and the
+hash value is used to shuffle the queue indices and deal a hand of
+size H, as follows.  We use a hash function that produces at least 64
+bits, and 64 of those bits are taken as an unsigned integer we will
+call V.  The next step is finding the unique set of integers A[0] in
+[0, numQueues), A[1] in [0, numQueues-1), … A[H-1] in
+[0, numQueues-(H-1)), A[H] >= 0 such that V = sum[i=0, 1, ...H] A[i] *
+ff(numQueues, i), where ff(N, M) is the falling factorial N!/(N-M)!.
+The probability distributions of each of these A’s will not be
+perfectly even, but we constrain the configuration such that
+ff(numQueues, H) is less than 2^60 to keep the unevenness small.  Then
+the coefficients A[0], … A[H-1] are converted into queue indices I[0],
+… I[H-1] as follows.  I[0] = A[0].  I[1] is the A[1]’th entry in the
+list of queue indices excluding I[0].  I[2] is the A[2]’th entry in
+the list of queue indices excluding I[0] and I[1].  And so on.
+
+The lengths of the queues identified by I[0], I[1], … I[H-1] are
+examined, and the request is put in one of the shortest queues.
+
+For example, if a RequestPriority has numQueues=128 and handSize=6,
+the hash value V is converted into 6 unique queue indices plus
+3905000064000*A[6].  There are 128 choose 6, which is about 5.4
+billion, sets of 6 integers in the range [0,127].  Thus, if there is
+one heavy flow and many light flows, the probability of a given light
+flow hashing to the same set of 6 queues as the heavy flow is about
+one in 5.4 billion.
+
+It is the queues that compete fairly.
```

Comment:
We should probably explicitly point out the corollary, which is that if the distinguisher method doesn't match the commonalities in the elephant's traffic, then the elephant can stop on all queues simultaneously. E.g., if the distinguisher method is by username, but actually the elephant is e.g. a sidecar which makes requests using its host pod's service account, then this evades the fairness.

I have a somewhat hairbrained scheme to address this, I don't have time to clean it up right now and I think we can live with this in the first pass anyway. Sysadmins, use correct distinguisher methods!